keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt

```diff
@@ -0,0 +1,149 @@
+import com.netflix.spinnaker.keel.notifications.NotificationScope.RESOURCE
+import com.netflix.spinnaker.keel.notifications.Notifier.UNHEALTHY
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.persistence.NoSuchResourceException
+import com.netflix.spinnaker.keel.persistence.NotifierRepository
+import com.netflix.spinnaker.keel.persistence.OrphanedResourceException
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import org.springframework.context.annotation.Configuration
+import org.springframework.context.event.EventListener
+import org.springframework.core.env.Environment
+import org.springframework.stereotype.Component
+
+/**
+ * A class that forwards notifications for resources based on the environment
+ * notification config.
+ *
+ * This class receives every notification, but notifies only once per
+ *  notify.waiting-duration (default once per day) if the notification
+ *  is a duplicate
+ *
+ * Note: this implementation is simple functionality. We could add more advanced
+ *  functionality by allowing the message to contain whether we should notify
+ *  more than once (and the frequency) and saving that into the database.
+ *  We could save the notification and check the database periodically to see
+ *  what needs to be sent, instead of relying on the events to keep coming.
+ *  We could allow the event to contain the full echo notification instead of
+ *  building it in this class, which would allow buttons and things in some
+ *  messages.
+ */
+@Configuration
+@EnableConfigurationProperties(KeelNotificationConfig::class)
+@Component
+class ResourceNotifier(
+  private val echoService: EchoService,
+  private val keelRepository: KeelRepository,
+  private val notifierRepository: NotifierRepository,
+  private val keelNotificationConfig: KeelNotificationConfig,
+  private val springEnv: Environment,
+  private val spectator: Registry
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+  private val NOTIFICATION_SENT_ID = "keel.notification.sent"
+
+  private val notificationsEnabled: Boolean
+    get() = springEnv.getProperty("keel.notifications.resource", Boolean::class.java, true)
+
+  @EventListener(ResourceNotificationEvent::class)
+  fun onResourceNotificationEvent(event: ResourceNotificationEvent) {
+    if (notificationsEnabled){
+      val shouldNotify = notifierRepository.addNotification(event.scope, event.identifier, event.notifier)
+      if (shouldNotify) {
+        notify(event)
+        notifierRepository.markSent(event.scope, event.identifier, event.notifier)
+        spectator.counter(
+          NOTIFICATION_SENT_ID,
+          listOf(BasicTag("notifier", event.notifier.name))
+        )
+          .runCatching { increment() }
+          .onFailure {
+            log.error("Exception incrementing {} counter: {}", NOTIFICATION_SENT_ID, it.message)
```

Comment:
I recommend passing the exception itself (`it`) instead of just the message (`it.message`) as the last argument to `log.error`. That way, both the message and the stack trace will get captured to the log.